File: ts-src/src/tools/read-file.ts
import { readFile } from 'node:fs/promises'
import { z } from 'zod'
import type { ToolDefinition } from '../tool.js'
import { resolveToolPath } from '../workspace.js'

type Input = {
  path: string
  offset?: number
  limit?: number
}

const DEFAULT_READ_LIMIT = 8000
const MAX_READ_LIMIT = 20000

export const readFileTool: ToolDefinition<Input> = {
  name: 'read_file',
  description:
    'Read a UTF-8 text file relative to the workspace root. Large files can be read in chunks via offset and limit.',
  inputSchema: {
    type: 'object',
    properties: {
      path: { type: 'string' },
      offset: { type: 'number' },
      limit: { type: 'number' },
    },
    required: ['path'],
  },
  schema: z.object({
    path: z.string(),
    offset: z.number().int().min(0).optional(),
    limit: z.number().int().min(1).max(MAX_READ_LIMIT).optional(),
  }),
  async run(input, context) {
    const target = await resolveToolPath(context, input.path, 'read')
    const content = await readFile(target, 'utf8')
    const offset = Math.max(0, input.offset ?? 0)
    const limit = Math.min(MAX_READ_LIMIT, input.limit ?? DEFAULT_READ_LIMIT)
    const end = Math.min(content.length, offset + limit)
    const chunk = content.slice(offset, end)
    const truncated = end < content.length
    const header = [
      `FILE: ${input.path}`,
      `OFFSET: ${offset}`,
      `END: ${end}`,
      `TOTAL_CHARS: ${content.length}`,
      truncated
        ? `TRUNCATED: yes - call read_file again with offset ${end}`
        : 'TRUNCATED: no',
      '',
    ].join('\n')

    return {
      ok: true,
      output: header + chunk,
    }
  },
}

File: py-src/minicode/tools/read_file.py
from __future__ import annotations

import time
from functools import lru_cache
from pathlib import Path

from minicode.tooling import ToolDefinition, ToolResult
from minicode.workspace import resolve_tool_path

DEFAULT_READ_LIMIT = 8000
MAX_READ_LIMIT = 20000


@lru_cache(maxsize=128)
def _get_cached_file_content(cache_key: tuple[str, float]) -> tuple[str, float]:
    """读取文件内容并返回（内容，读取时间）。

    使用 lru_cache 自动管理缓存，键为 (文件路径, 修改时间)。
    当文件修改时间变化时，缓存键变化，自动失效旧缓存。
    """
    file_path = Path(cache_key[0])
    content = file_path.read_text(encoding="utf-8")
    return content, time.monotonic()


def _get_file_content(target: Path) -> str:
    """获取文件内容，使用 @lru_cache 避免重复读取。"""
    try:
        stat = target.stat()
        mtime = stat.st_mtime
        cache_key = (str(target), mtime)
        content, _ = _get_cached_file_content(cache_key)
        return content
    except OSError:
        return target.read_text(encoding="utf-8")


def _validate(input_data: dict) -> dict:
    path = input_data.get("path")
    if not isinstance(path, str) or not path:
        raise ValueError("path is required")
    offset = int(input_data.get("offset", 0))
    limit = int(input_data.get("limit", DEFAULT_READ_LIMIT))
    if offset < 0:
        raise ValueError("offset must be >= 0")
    if limit < 1 or limit > MAX_READ_LIMIT:
        raise ValueError(f"limit must be between 1 and {MAX_READ_LIMIT}")
    return {"path": path, "offset": offset, "limit": limit}


def _run(input_data: dict, context) -> ToolResult:
    target = resolve_tool_path(context, input_data["path"], "read")

    try:
        content = _get_file_content(target)
    except UnicodeDecodeError:
        return ToolResult(
            ok=False,
            output=f"File {input_data['path']} appears to be binary. Cannot read as text.",
        )

    offset = input_data["offset"]
    limit = input_data["limit"]
    end = min(len(content), offset + limit)
    chunk = content[offset:end]
    truncated = end < len(content)
    header = "\n".join(
        [
            f"FILE: {input_data['path']}",
            f"OFFSET: {offset}",
            f"END: {end}",
            f"TOTAL_CHARS: {len(content)}",
            f"TRUNCATED: {'yes - call read_file again with offset ' + str(end) if truncated else 'no'}",
            "",
        ]
    )
    return ToolResult(ok=True, output=header + chunk)


read_file_tool = ToolDefinition(
    name="read_file",
    description="Read a UTF-8 text file relative to the workspace root.",
    input_schema={"type": "object", "properties": {"path": {"type": "string"}, "offset": {"type": "number"}, "limit": {"type": "number"}}, "required": ["path"]},
    validator=_validate,
    run=_run,
)
